File: frameworks/oxpecker/Handlers.fs
module HttpArena.Handlers

open System
open System.Globalization
open System.IO
open System.Text
open HttpArena.Services
open Microsoft.AspNetCore.Http
open Oxpecker

/// Reads an int query parameter through Oxpecker's query accessor, falling
/// back to `fallback` when the parameter is absent or unparsable.
let private queryInt (ctx: HttpContext) (key: string) (fallback: int) =
    match ctx.TryGetQueryValue key with
    | Some raw ->
        match Int32.TryParse(raw, NumberStyles.Integer, CultureInfo.InvariantCulture) with
        | true, value -> value
        | _ -> fallback
    | None -> fallback

let private queryFloat (ctx: HttpContext) (key: string) (fallback: float) =
    match ctx.TryGetQueryValue key with
    | Some raw ->
        match Double.TryParse(raw, NumberStyles.Float, CultureInfo.InvariantCulture) with
        | true, value -> value
        | _ -> fallback
    | None -> fallback

// ── Connection profiles ────────────────────────────────────────────────────

let pipeline: EndpointHandler = text "ok"

/// GET /baseline11 and GET /baseline2 — sum of the two query parameters.
let baseline: EndpointHandler =
    fun ctx ->
        let a = queryInt ctx "a" 0
        let b = queryInt ctx "b" 0
        ctx.WriteText(string (a + b))

/// POST /baseline11 — sum of the two query parameters plus the request body.
let baselineWithBody: EndpointHandler =
    fun ctx ->
        let a = queryInt ctx "a" 0
        let b = queryInt ctx "b" 0

        task {
            use reader = new StreamReader(ctx.Request.Body)
            let! body = reader.ReadToEndAsync()

            let fromBody =
                match Int32.TryParse(body, NumberStyles.Integer, CultureInfo.InvariantCulture) with
                | true, value -> value
                | _ -> 0

            return! ctx.WriteText(string (a + b + fromBody))
        }

// ── Workload profiles ──────────────────────────────────────────────────────

/// POST /echo — the request body written back verbatim.
///
/// The body is copied into a MemoryStream rather than piped straight to the
/// response because the length has to be known before the first byte goes out:
/// a chunked request carries no Content-Length to forward, and WriteBytes sets
/// one from the array it is given — zero included, for an empty body.
let echo: EndpointHandler =
    fun ctx ->
        task {
            use body = new MemoryStream()
            do! ctx.Request.Body.CopyToAsync body

            ctx.SetContentType "application/octet-stream"
            return! ctx.WriteBytes(body.ToArray())
        }

/// GET /json/{count}?m=N — the first `count` dataset items, each carrying a
/// total computed from the multiplier.
let json (count: int) : EndpointHandler =
    fun ctx ->
        let multiplier = queryInt ctx "m" 1
        let response = Dataset.getItems count multiplier
        ctx.WriteJsonChunked response

// ── Database profiles ──────────────────────────────────────────────────────

/// GET /async-db — Postgres range query over the unindexed price column.
let asyncDb: EndpointHandler =
    fun ctx ->
        let minPrice = queryFloat ctx "min" 10.0
        let maxPrice = queryFloat ctx "max" 50.0
        let limit = queryInt ctx "limit" 50
        task {
            let! response = Items.query minPrice maxPrice limit
            return! ctx.WriteJsonChunked response
        }   

/// GET /crud/items — paginated list by category.
let crudList: EndpointHandler =
    fun ctx ->
        let category = ctx.TryGetQueryValue "category" |> Option.defaultValue ""
        let page = queryInt ctx "page" 0
        let limit = queryInt ctx "limit" 0
        task {
            let! response = Items.list category page limit
            return! ctx.WriteJsonChunked response
        }

/// GET /crud/items/{id} — cache-aside single-item read, reporting the cache
/// outcome through X-Cache.
let crudRead (id: int) : EndpointHandler =
    fun ctx ->
        task {
            match! Items.read id with
            | ValueNone ->
                ctx.SetStatusCode 404
            | ValueSome result ->
                ctx.SetHttpHeader("X-Cache", if result.CacheHit then "HIT" else "MISS")
                match result.Value with
                | TypedItem item ->
                    return! ctx.WriteJsonChunked item
                | SerializedItem cached ->
                    // Already JSON on the Redis path — write the cached bytes
                    // back rather than round-tripping them through the serializer.
                    ctx.SetContentType "application/json"
                    return! ctx.WriteBytes(Encoding.UTF8.GetBytes cached)
        }

/// POST /crud/items — create (upsert on id conflict).
let crudCreate: EndpointHandler =
    fun ctx ->
        task {
            let! input = ctx.BindJson<CrudItemInput>()
            let! created = Items.create input
            ctx.SetStatusCode 201
            return! ctx.WriteJsonChunked created
        }

/// PUT /crud/items/{id} — update and invalidate the cached entry.
let crudUpdate (id: int) : EndpointHandler =
    fun ctx ->
        task {
            let! input = ctx.BindJson<CrudItemInput>()
            match! Items.update id input with
            | None ->
                ctx.SetStatusCode 404
            | Some updated ->
                return! ctx.WriteJsonChunked updated
        }
            

// ── Template profile ───────────────────────────────────────────────────────

/// GET /fortunes — DB query plus an Oxpecker.ViewEngine render.
let fortunes: EndpointHandler =
    fun ctx ->
        task {
            let! rows = Fortunes.getRows ()
            return! ctx.WriteHtmlViewChunked(Views.fortunes rows)
        }

File: frameworks/carter/Handlers.cs
using System.Text;
using System.Text.Json;

using Microsoft.AspNetCore.Http;

namespace HttpArena.Carter;

public static class Handlers
{
    private static readonly JsonSerializerOptions JsonOpts = new(JsonSerializerDefaults.Web);

    /// Sum of every query parameter whose value parses as an integer, plus the
    /// body when there is one. A non-numeric value is skipped rather than
    /// failing the request.
    public static async Task Baseline11(HttpContext ctx)
    {
        long total = 0;
        foreach (var pair in ctx.Request.Query)
        {
            if (long.TryParse(pair.Value.ToString(), out var n)) total += n;
        }

        if (ctx.Request.ContentLength is > 0 || ctx.Request.Headers.TransferEncoding.Count > 0)
        {
            using var reader = new StreamReader(ctx.Request.Body, Encoding.UTF8);
            var body = await reader.ReadToEndAsync();
            if (long.TryParse(body.Trim(), out var b)) total += b;
        }

        ctx.Response.ContentType = "text/plain";
        await ctx.Response.WriteAsync(total.ToString());
    }

    public static async Task JsonItems(HttpContext ctx, DatasetService dataset, int count)
    {
        long m = 1;
        if (ctx.Request.Query.TryGetValue("m", out var raw) && long.TryParse(raw.ToString(), out var parsed))
            m = parsed;

        var all = dataset.Items;
        var n = Math.Min(count, all.Length);
        var items = new List<OutItem>(n);
        for (var i = 0; i < n; i++)
        {
            var d = all[i];
            items.Add(new OutItem
            {
                Id = d.Id,
                Name = d.Name,
                Category = d.Category,
                Price = d.Price,
                Quantity = d.Quantity,
                Active = d.Active,
                Tags = d.Tags,
                Rating = d.Rating,
                Total = d.Price * d.Quantity * m,
            });
        }

        ctx.Response.ContentType = "application/json";
        await JsonSerializer.SerializeAsync(ctx.Response.Body, new OutList { Items = items, Count = n }, JsonOpts);
    }

    /// Echo: the bytes that arrived go back unchanged. Collected first because
    /// the response needs a Content-Length, and a chunked request carries none
    /// to forward until the body is in.
    public static async Task EchoBody(HttpContext ctx)
    {
        using var ms = new MemoryStream();
        await ctx.Request.Body.CopyToAsync(ms);
        var body = ms.GetBuffer().AsMemory(0, (int)ms.Length);
        ctx.Response.ContentType = "application/octet-stream";
        ctx.Response.ContentLength = body.Length;
        await ctx.Response.Body.WriteAsync(body);
    }
}
